### PR TITLE
convert ra-ui-materialui layout to typescript

### PR DESCRIPTION
**Pull request by MohammedFaragallah**
refs #4505

### CODE REVIEW

**Comment by MohammedFaragallah** [comment]
@fzaninotto this is ready for review but if it's not clear enough I'll be able to add some comments soon

some questions
- I think we should create a separate ErrorBounday component and covert layout to a functional component
- some component destructure `classes` property out I think we can remove this pattern and typescript will warn users if they pass a non-supported property

**Comment by fzaninotto** [comment]
> I think we should create a separate ErrorBounday component and convert Layout to a functional component

I agree, good idea!

> some component destructure classes property out I think we can remove this pattern and typescript will warn users if they pass a non-supported property

Not sure I understand. Can you give me an example?

File: packages/ra-ui-materialui/src/defaultTheme.ts [hunk 1]
```diff
@@ -7,11 +27,6 @@ export default {
             contrastText: '#fff',
         },
     },
-    typography: {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/defaultTheme.ts:10]: Why did you remove this one?

File: packages/ra-ui-materialui/src/layout/CardContentInner.tsx [hunk 2]
```diff
@@ -30,11 +30,11 @@ const useStyles = makeStyles(
  * padding double the spacing between each CardContent, leading to too much
  * wasted space. Use this component as a CardContent alternative.
  */
-const CardContentInner = props => {
-    const { className, children } = props;
+const CardContentInner: FC<CardContentProps> = props => {
+    const { className, children, ...rest } = props;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/CardContentInner.tsx:34]: Not necessary for now, may have side effects... Please revert this change.

File: packages/ra-ui-materialui/src/layout/Error.tsx [hunk 3]
```diff
@@ -52,7 +52,15 @@ function goBack() {
     window.history.go(-1);
 }
 
-const Error = props => {
+export interface ErrorComponentProps {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Error.tsx:55]: ErrocComponentProps => ErrorProps (we never include the name "component" in other interface names)

File: packages/ra-ui-materialui/src/layout/Error.tsx [hunk 4]
```diff
@@ -86,12 +94,8 @@ const Error = props => {
                     </ExpansionPanel>
                 )}
                 <div className={classes.toolbar}>
-                    <Button
-                        variant="contained"
-                        icon={<History />}
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Error.tsx:91]: I think `startIcon` is the right prop to get the alignment icon/text

File: packages/ra-ui-materialui/src/layout/Layout.tsx [hunk 5]
```diff
@@ -74,11 +85,59 @@ const sanitizeRestProps = ({
...
+    className?: string;
+    customRoutes?: any[];
+    dashboard?: DashboardComponent;
+    error?: ComponentType<ErrorComponentProps>;
+    logout?: ReactNode;
+    menu?: ComponentType<MenuProps>;
+    notification?: ComponentType<NotificationProps>;
+    open?: boolean;
+    sidebar?: ComponentType<DrawerProps>;
+    title: TitleComponent;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Layout.tsx:105]: It's optional at build time (but required at runtime).

File: packages/ra-ui-materialui/src/layout/Loading.tsx [hunk 6]
```diff
@@ -53,7 +59,6 @@ const Loading = props => {
 };
 
 Loading.propTypes = {
-    classes: PropTypes.object,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Loading.tsx:56]: You shouldn't remove this one. When `classes` is present in the props, useStyles() uses it to override the default styles.

File: packages/ra-ui-materialui/src/layout/Loading.tsx [hunk 7]
```diff
@@ -33,7 +33,13 @@ const useStyles = makeStyles(
     { name: 'RaLoading' }
 );
 
-const Loading = props => {
+interface Props {
+    className?: string;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Loading.tsx:37]: you miss one prop: `classes?`

File: packages/ra-ui-materialui/src/layout/Menu.tsx [hunk 8]
```diff
@@ -1,4 +1,4 @@
-import React, { FC, ReactElement } from 'react';
+import React, { FC, ReactElement, ReactNode } from 'react';
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Menu.tsx:1]: Then I think ReactElement is not used anymore

File: packages/ra-ui-materialui/src/layout/NotFound.tsx [hunk 9]
```diff
@@ -45,8 +46,14 @@ function goBack() {
     window.history.go(-1);
 }
 
-const NotFound = props => {
-    const { className, classes: classesOverride, title, ...rest } = props;
+interface Props extends RouteComponentProps {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/NotFound.tsx:49]: make sure you export it.

File: packages/ra-ui-materialui/src/layout/NotFound.tsx [hunk 10]
```diff
@@ -45,8 +46,14 @@ function goBack() {
...
+    className: string;
+    title: TitleProps['defaultTitle'];
+    classes?: object;
+}
+
+const NotFound: FC<Props> = props => {
+    const { className, title, ...rest } = props;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/NotFound.tsx:56]: you should also add `classes: classesOverride` to the destructuring, or the root `<div>` may end up with a `classes` attribute which isn't DOM compliant.

File: packages/ra-ui-materialui/src/layout/NotFound.tsx [hunk 11]
```diff
@@ -45,8 +46,14 @@ function goBack() {
     window.history.go(-1);
 }
 
-const NotFound = props => {
-    const { className, classes: classesOverride, title, ...rest } = props;
+interface Props extends RouteComponentProps {
+    className: string;
+    title: TitleProps['defaultTitle'];
+    classes?: object;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/NotFound.tsx:52]: I think MUI offers a way to infer the accepted type based on the makeStyles.

File: packages/ra-ui-materialui/src/layout/NotFound.tsx [hunk 12]
```diff
@@ -62,8 +69,8 @@ const NotFound = props => {
                 <div>{translate('ra.message.not_found')}.</div>
             </div>
             <div className={classes.toolbar}>
-                <Button variant="contained" icon={<History />} onClick={goBack}>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/NotFound.tsx:65]: use startIcon

File: packages/ra-ui-materialui/src/layout/Sidebar.tsx [hunk 13]
```diff
@@ -45,22 +55,21 @@ const useStyles = makeStyles(
...
 
-const Sidebar = props => {
-    const {
-        children,
-        closedSize,
-        size,
-        classes: classesOverride,
-        ...rest
-    } = props;
+interface Props extends DrawerProps {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Sidebar.tsx:58]: let's call it `SidebarProps` and export it.

File: packages/ra-ui-materialui/src/layout/Sidebar.tsx [hunk 14]
```diff
@@ -45,22 +55,21 @@ const useStyles = makeStyles(
...
-    useSelector(state => state.locale); // force redraw on locale change
+    const isXSmall = useMediaQuery<Theme>(theme =>
+        theme.breakpoints.down('xs')
+    );
+    const isSmall = useMediaQuery<Theme>(theme => theme.breakpoints.down('sm'));
+    const open = useSelector((state: ReduxState) => state.admin.ui.sidebarOpen);
     const handleClose = () => dispatch(setSidebarVisibility(false));
     const toggleSidebar = () => dispatch(setSidebarVisibility(!open));
-    const classes = useStyles({ ...props, open });
+    const classes = useStyles({ open });
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Sidebar.tsx:72]: not good, we must pass the complete `props` object or this will cause rerenderings

File: packages/ra-ui-materialui/src/layout/Sidebar.tsx [hunk 15]
```diff
@@ -45,22 +55,21 @@ const useStyles = makeStyles(
...
-const Sidebar = props => {
-    const {
-        children,
-        closedSize,
-        size,
-        classes: classesOverride,
-        ...rest
-    } = props;
+interface Props extends DrawerProps {
+    children: ReactElement;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Sidebar.tsx:59]: I think it should be `React.ReactNode`

File: packages/ra-ui-materialui/src/layout/TitleForRecord.tsx [hunk 16]
```diff
@@ -0,0 +1,18 @@
+import React, { FC } from 'react';
+import PropTypes from 'prop-types';
+import Title, { TitlePropType, TitleProps } from './Title';
+
+interface Props extends TitleProps {
+    record: TitleProps['record'];
+}
+
+const TitleForRecord: FC<Props> = props =>
+    props.record ? <Title {...props} /> : null;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/TitleForRecord.tsx:10]: the previous version white listed the props it passed to the `<Title>` component. Please keep that behavior.

File: packages/ra-ui-materialui/src/defaultTheme.ts (hunk 1 above)
Comment (MohammedFaragallah) [reply packages/ra-ui-materialui/src/defaultTheme.ts:10]: I didn't mean to push that with this PR but I checked everywhere there is no use for variant `title`

File: packages/ra-ui-materialui/src/layout/Sidebar.tsx (hunk 15 above)
Comment (MohammedFaragallah) [reply packages/ra-ui-materialui/src/layout/Sidebar.tsx:59]: ReactNode gives an error

`Type 'string' is not assignable to type 'ReactElement<any, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)>'.`

File: packages/ra-ui-materialui/src/layout/NotFound.tsx (hunk 11 above)
Comment (MohammedFaragallah) [reply packages/ra-ui-materialui/src/layout/NotFound.tsx:52]: I don't know about it but something like ReturnType<typeof useStyles> should work?

File: packages/ra-ui-materialui/src/layout/NotFound.tsx (hunk 10 above)
Comment (MohammedFaragallah) [reply packages/ra-ui-materialui/src/layout/NotFound.tsx]: > > some component destructure classes property out I think we can remove this pattern and typescript will warn users if they pass a non-supported property
> 
> Not sure I understand. Can you give me an example?

that's what I was talking about, didn't know about `makeStyles` override behavior, Thank you

File: packages/ra-ui-materialui/src/layout/NotFound.tsx (hunk 11 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/layout/NotFound.tsx]: absolutely!

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/layout/NotFound.tsx]: No, wait, it won't work. the `classes` *argument* passed to `useStyles` isn't the same type as the *result* of `useStyles`, which is usually named `classes`.

I'm looking for the right way to type the `classes` override (the material-ui documentation doesn't mention it).

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/layout/NotFound.tsx]: Okay, there is no easy way to type these - see my comment in https://github.com/mui-org/material-ui/issues/17973#issuecomment-638868260.

Let's default to `any` for now.

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/layout/NotFound.tsx]: Or we can add a generic type to `ra-ui-material-ui`:

```tsx
import { CSSProperties, CreateCSSProperties } from "@material-ui/styles";

type PropsFunc<Props extends object, T> = (props: Props) => T;

export type ClassesOverride<
  UseStyles extends (props: any) => Record<string, string>,
  Props extends object = {}
> = Partial<
  {
    [p in keyof ReturnType<UseStyles>]:
      | CSSProperties
      // JSS property bag where values are based on props
      | CreateCSSProperties<Props>
      // JSS property bag based on props
      | PropsFunc<Props, CreateCSSProperties<Props>>
  }
>;
```

which allows to type our `classes` props as follows:

```tsx
interface Props {
    classes: ClassesOverride<typeof useStyles>;
    ...
}
```

What do you think?

Comment (MohammedFaragallah) [reply packages/ra-ui-materialui/src/layout/NotFound.tsx]: Excellent but from what I understand the parent component passes a `classes` object each key is a `className` string, not `CSSProperties` object?

```jsx
import { makeStyles } from '@material-ui/core';
import React, { FC } from 'react';

const useNotFoundStyles = makeStyles(
	theme => ({
		root: {
			display: 'flex',
		},
	}),
	{ name: 'RaNotFound' },
);

type ClassesOverride<
	UseStyles extends (props: any) => Record<string, string>
> = Partial<Record<keyof ReturnType<UseStyles>, string>>;

interface NotFoundProps {
	classes?: ClassesOverride<typeof useNotFoundStyles>;
}

export const NotFound: FC<NotFoundProps> = props => {
	const classes = useNotFoundStyles(props);
	return <div className={classes.root}>DEMO</div>;
};
```

```jsx
const useLayoutStyles = makeStyles(
  theme => ({
    root: {
      display: "block"
    }
  }),
  { name: "RaLayout" }
);

interface LayoutProps {}

export const Layout: React.FC<LayoutProps> = props => {
  const classes = useLayoutStyles(props);
  return (
    <div>
      <NotFound classes={{ root: classes.root }} />
    </div>
  );
};
```

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/layout/NotFound.tsx]: Totally, you're right and I was too tired when writing my comment. Sorry for that!

File: packages/ra-ui-materialui/src/layout/Title.tsx [hunk 17]
```diff
@@ -26,6 +32,13 @@ const Title = ({ className, defaultTitle, locale, record, title, ...rest }) => {
     return createPortal(titleElement, container);
 };
 
+export interface TitleProps {
+    defaultTitle: TitleComponent;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Title.tsx:36]: it's optional. Either one of `title` or `defaultTitle` should be defined.

File: packages/ra-ui-materialui/src/layout/TitleForRecord.tsx [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+        <Title
+            title={title}
+            record={record}
+            defaultTitle={defaultTitle}
+            className={className}
+        />
+    ) : null;
+
+interface Props extends TitleProps {
+    record: TitleProps['record'];
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/TitleForRecord.tsx:21]: I don't understand this type. It specializes nothing.

File: packages/ra-ui-materialui/src/layout/NotFound.tsx [hunk 19]
```diff
@@ -76,13 +81,21 @@ const sanitizeRestProps = ({
...
     ...rest
-}) => rest;
+}: Partial<NotFoundProps>): Omit<
+    Partial<NotFoundProps>,
+    keyof RouteComponentProps | 'className' | 'title'
+> => rest;
+
+export interface NotFoundProps extends RouteComponentProps {
+    className: string;
+    title: TitleProps['defaultTitle'];
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/NotFound.tsx:91]: Optional I think

File: packages/ra-ui-materialui/src/layout/NotFound.tsx (hunk 11 above)
Comment (MohammedFaragallah) [reply packages/ra-ui-materialui/src/layout/NotFound.tsx]: No worries you are working on so many things, I'll add this type to Layout components for now and start a new PR for all `ra-material-ui` other components.

File: packages/ra-ui-materialui/src/layout/TitleForRecord.tsx (hunk 18 above)
Comment (MohammedFaragallah) [reply packages/ra-ui-materialui/src/layout/TitleForRecord.tsx:21]: to make `record` required, but to make things more clear I can add a helper type?

```ts
export type SomeRequired<T, U extends keyof T> = Omit<T, U> &
  Required<Pick<T, U>>;
```

```diff
+ interface Props extends <TitleProps,'record'> {
- interface Props extends TitleProps {
-     record: TitleProps['record'];
```

**Comment by WiXSL** [comment]
Is this PR still valid?

**Comment by Luwangel** [comment]
@WiXSL I don't know. Let's ask François, he'll be back soon :wink: 

@fzaninotto I'm not sure I understand why you closed that PR.
